### PR TITLE
Do not try to enable AppArmor in rootless mode

### PR DESCRIPTION
**Pull request by marcov**
When in rootless mode it's not possible to load profiles or check which profiles are loaded. 

I am proposing this change is to avoid users having to specify `--security-opt apparmor=unconfined` every time they are running rootless.

If this is something that makes sense I'll also add some tests.

Actually a better way would be to get the parent process effective capabilities to see whether CAP_MAC_ADMIN is set. CAP_MAC_ADMIN allows controlling AppArmor without root permissions. 
However I am not sure how to do it given that at the check time we are root in the user namespace and calling capget will give full effective capabilities.

Something like:
```diff
-       aaControllable := !rootless.IsRootless()
+       aaControllable := true
+
+       if rootless.IsRootless() {
+               myCaps, err := capability.NewPid2(0)
+               if err != nil {
+                       return fmt.Errorf("Error getting current process capabilities")
+               }
+               err = myCaps.Load()                                                                                                 
+               err != nil {
+                       return fmt.Errorf("Error getting current process capabilities")
+               }
+               aaControllable = myCaps.Get(capability.EFFECTIVE, capability.CAP_MAC_ADMIN)
+       }
```

### CODE REVIEW

**Comment by marcov** [comment]
@vrothberg PTAL

**Comment by vrothberg** [comment]
The proposed changes make the code really complex. I think it's sufficient to add another check here https://github.com/projectatomic/libpod/blob/master/cmd/podman/create.go#L235 and here https://github.com/projectatomic/libpod/blob/master/cmd/podman/create.go#L199. 

But I think it's timeo put the AppArmor loading into a speparate function `func loadAAProfile(profile string) error` (for instance) which would allow us to do a simple `return nil` instead of increasingly complex branches. For the rootless case, I suggest to log that Podman is not loading the AppArmor profile because of the above mentioned reasons.

**Comment by vrothberg** [comment]
@giuseppe PTAL

**Comment by marcov** [comment]
Yes I agree that having a function is probably the cleanest way to do it. I'll do a rewrite.

**Comment by giuseppe** [comment]
I agree with @vrothberg, it is better if you could refactor it in a function.

If it is something useful to have, we could add the check for the effective capabilities before doing the re-exec and store that result into an env variable.

If the re-exec code will grow to be too complex we might need to use some other data format instead of env variables, but it is not something for this PR.

**Comment by rhatdan** [comment]
@marcov I added the WIP to make sure no one accidently merges it,  Remove it when you are ready to have the PR reviewed.

**Comment by marcov** [comment]
@giuseppe thanks for the hint.
I realized we need to check ambient capabilities (permitted & inheritable) instead of effective.
Also it would be best to store the full capability bitmask in the environment variable, but the only way I found to extract the full 64-bit mask is via procfs. So I thought about using [psgo](https://github.com/containers/psgo) for this.

**Comment by giuseppe** [comment]
is it something we could do using directly `prctl` on the current thread instead of going through `/proc`?

**Comment by vrothberg** [comment]
> is it something we could do using directly prctl on the current thread instead of going through /proc?

I think that's possible. There's already quite some logic in `psgo/internal/capabilities`. We could sacrifice this package and make it public by moving it to `psgo/pkg/capabilities` and add a new API to do that. @marcov, what do you think?

**Comment by marcov** [comment]
@vrothberg: I like your idea, but consider there's already a [gocapability](https://github.com/syndtr/gocapability/) package to operate on capabilities (even though this does not give access to the full capability bitmasks)

@giuseppe: `prctl` can only read one capabilty bit per call. I thought that reading the full-mask would be more future proof if we need to get other bits in future.

**Comment by mheon** [comment]
bot, add author to whitelist

**Comment by vrothberg** [comment]
@marcov Then I suggest going with your initial proposal of extending `psgo` with the required API and go from there.

**Comment by giuseppe** [comment]
bot, add author to whitelist

**Comment by giuseppe** [comment]
yes extending psgo is a good idea, but in our case don't we know exactly what capabilities are needed?

**Comment by marcov** [comment]
`prctl` is enough for now but we'll need to call it n times if in future we have to check for n capabilities. That's why I proposed reading the full capabilities once only with psgo. Anyway I am fine in both ways.

**Comment by marcov** [comment]
I actually need to correct myself regarding what is possible to do with AppArmor as non-root user.

The pkg `apparmor` uses `apparmor_parser` as a frontend to AppArmor, and that requires root privileges.

I am sorry for having provided you wrong information.
So for now I'll avoid checking capabilities and just do the refactoring of AppArmor code into a function as per your suggestion.

PS: some more details after checking in the AppArmor kernel code (these are my conclusions and may still be imprecise):
- reading `/sys/kernel/security/apparmor/profiles`  requires to be root user (see `policy_view_capable()` in `policy.c`).
- administering AppArmor requires to be root user **and** CAP_MAC_ADMIN (see `policy_admin_capable()` in `policy.c`).

File: cmd/podman/create.go [hunk 1]
```diff
@@ -250,6 +204,70 @@ func parseSecurityOpt(config *cc.CreateConfig, securityOpts []string) error {
...
+				config.ApparmorProfile = con[1]
+			case "seccomp":
+				config.SeccompProfilePath = con[1]
+			default:
+				return fmt.Errorf("Invalid --security-opt 2: %q", opt)
+			}
+		}
+	}
+
+	if rootless.IsRootless() {
```
Comment (vrothberg) [inline cmd/podman/create.go]: I would prefer this to also go into `loadAAProfile` to have only a single function call + error check. Would you mind renaming the function to `loadAppArmor`? `AA` isn't very obvious for readers.

**Comment by TomSweeneyRedHat** [comment]
changes LGTM, but would like head nods form @vrothberg and @giuseppe

**Comment by vrothberg** [comment]
bot, retest this please

File: cmd/podman/create.go [hunk 2]
```diff
@@ -148,52 +149,17 @@ func createCmd(c *cli.Context) error {
...
-		err       error
-	)
+func loadAppArmor(config *cc.CreateConfig) error {
+	if rootless.IsRootless() {
+		switch config.ApparmorProfile {
+		case "":
+			logrus.Warn("The AppArmor default profile will not be loaded in rootless mode")
+		case "unconfined":
+		default:
+			return fmt.Errorf("The specified AppArmor security profile cannot bet loaded in rootless mode")
```
Comment (vrothberg) [inline cmd/podman/create.go]: Please remove the blank line.

File: cmd/podman/create.go [hunk 3]
```diff
@@ -148,52 +149,17 @@ func createCmd(c *cli.Context) error {
 	return nil
 }
 
-func parseSecurityOpt(config *cc.CreateConfig, securityOpts []string) error {
-	var (
-		labelOpts []string
-		err       error
-	)
+func loadAppArmor(config *cc.CreateConfig) error {
```
Comment (vrothberg) [inline cmd/podman/create.go]: Can you add a comment/description to `loadAppArmor()`?

Comment (vrothberg) [reply cmd/podman/create.go]: One thing I noticed is that the format of logs and errors is a bit inconsistent in this function. Some start capital, some start lower case. Also `AppArmor profile` vs `AppArmor security profile`.

**Comment by TomSweeneyRedHat** [comment]
Note to self, add a little comment in the test for the whale says positioning.

**Comment by mheon** [comment]
@rh-atomic-bot r+

**Pull request closed without merging**